language/move-core/types/src/vm_status.rs

```diff
@@ -106,86 +107,83 @@ impl std::error::Error for VMStatus {}
 
 impl VMStatus {
     /// Create a new VM status with major status `major_status`.
-    pub fn new(major_status: StatusCode) -> Self {
+    pub fn new(major_status: StatusCode, sub_status: Option<u64>, message: Option<String>) -> Self {
```

Comment:
Could `message` be an `Option<Location>` instead of a String? Either that or a separate more structured type:
```
enum SourceContext {
    Script
    Module(ModuleId)
}
```
Reason being we want this type to be pretty well structured since it will be a part of consensus (and therefore they all have to agree on the exact status), and also stored in the transaction accumulator.

Additionally, if the `message` is `None` we should also set the `sub_status` to `None` I think. We only want to put out source-generated errors otherwise the versioning story for `sub_status` becomes more complex since it is then versioned by the VM and modules+txn scripts as opposed to just the modules+txn scripts.


Edit: on second thought, it might be easier if I make this change as part of the PR that adds VMStatus to the rest of the system. If that makes more sense to you I'm happy to make these changes in that PR.

Reply:
Let's talk after this lands about next steps for the errors